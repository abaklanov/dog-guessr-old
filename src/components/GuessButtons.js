import React from "react";
import Axios from "axios";
import { Button } from "react-bootstrap";

class GuessButtons extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      schema: [],
      areButtonsDisabled: false
    };
  }

  componentDidMount() {
    Axios.get("https://dog.ceo/api/breeds/list")
      .then(
        (resp => {
          this.setState({
            schema: this.getSchema(resp.data.message),
            areButtonsDisabled: false
          });
        }).bind(this)
      )
      .catch(err => {});
  }

  getRandomBreeds(breeds, num = 3) {
    const result = [];
    for (let i = 0; i < num; i++) {
      const randomBreed = this.getRandomBreed(breeds);
      result.push(randomBreed);
    }
    return result;
  }

  getRandomBreed(breeds) {
    const min = 0;
    const max = breeds.length;
    const rand = Math.floor(Math.random() * (max - min)) + min;
    return breeds[rand];
  }

  getArrayOfGuesses(breeds) {
    /**
     * Shuffles array in place. ES6 version
     * @param {Array} a items An array containing the items.
     */
    function shuffle(a) {
      for (let i = a.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [a[i], a[j]] = [a[j], a[i]];
      }
      return a;
    }

    // Clone
    const res = this.getRandomBreeds(breeds);
    res.push(this.props.breedToGuess);
    return shuffle(res);
  }

  getSchema(breeds) {
    const guesses = this.getArrayOfGuesses(breeds);
    return guesses.map(guess => {
      return {
        btnVariant: "light",
        isCorrect: guess === this.props.breedToGuess,
        answer: guess
      };
    });
  }

  doGuess(index) {
    const currentSchemaState = this.state.schema.slice(0);

    currentSchemaState[index].btnVariant = currentSchemaState[index].isCorrect
      ? "success"
      : "danger";

    this.setState({
      schema: currentSchemaState,
      areButtonsDisabled: true
    });

    setTimeout(() => {
      window.location.replace("./");
    }, 3000);
  }

  render() {
    const possibleGuesses = this.state.schema.map((element, index) => (
      <div className="row" key={index}>
        <div className="col-lg-12 text-center">
          <Button
            variant={element.btnVariant}
            className="guessBtn m-1"
            onClick={e => this.doGuess(index, e)}
            disabled={this.state.areButtonsDisabled ? "disabled" : ""}
          >
            {element.answer}
          </Button>
        </div>
      </div>
    ));
    return (
      <div className="row justify-content-center">
        <div className="col-lg-4">{possibleGuesses}</div>
      </div>
    );
  }
}

export default GuessButtons;
