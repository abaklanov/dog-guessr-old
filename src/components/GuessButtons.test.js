import React from "react";
import { shallow } from "enzyme";
import GuessButtons from "./GuessButtons";
import { Button } from "react-bootstrap";

const wrapper = shallow(<GuessButtons />);

jest.mock("axios", () => {
  return {
    get: () =>
      Promise.resolve({
        status: 200,
        data: {
          status: "success",
          message: [
            "affenpinscher",
            "african",
            "airedale",
            "akita",
            "appenzeller",
            "basenji",
            "beagle",
            "bluetick",
            "borzoi",
            "bouvier",
            "boxer",
            "brabancon",
            "briard",
            "bulldog",
            "bullterrier",
            "cairn",
            "cattledog",
            "chihuahua",
            "chow",
            "clumber",
            "cockapoo",
            "collie",
            "coonhound",
            "corgi",
            "cotondetulear",
            "dachshund",
            "dalmatian",
            "dane",
            "deerhound",
            "dhole",
            "dingo",
            "doberman",
            "elkhound",
            "entlebucher",
            "eskimo",
            "frise",
            "germanshepherd",
            "greyhound",
            "groenendael",
            "hound",
            "husky",
            "keeshond",
            "kelpie",
            "komondor",
            "kuvasz",
            "labrador",
            "leonberg",
            "lhasa",
            "malamute",
            "malinois",
            "maltese",
            "mastiff",
            "mexicanhairless",
            "mix",
            "mountain",
            "newfoundland",
            "otterhound",
            "papillon",
            "pekinese",
            "pembroke",
            "pinscher",
            "pointer",
            "pomeranian",
            "poodle",
            "pug",
            "puggle",
            "pyrenees",
            "redbone",
            "retriever",
            "ridgeback",
            "rottweiler",
            "saluki",
            "samoyed",
            "schipperke",
            "schnauzer",
            "setter",
            "sheepdog",
            "shiba",
            "shihtzu",
            "spaniel",
            "springer",
            "stbernard",
            "terrier",
            "vizsla",
            "weimaraner",
            "whippet",
            "wolfhound"
          ]
        }
      })
  };
});
it("renders correctly", () => {
  // TODO: Fix as breed name changes
  //expect(wrapper).toMatchSnapshot();
});

it("gets a breed being guessed as a prop", () => {
  // TODO: implement
});

it("stores 4 possible answers", () => {
  expect(wrapper.state().possibleGuesses).toHaveLength(4);
});

it("shows four buttons with breed names on it", () => {
  const guessButtons = wrapper.find(Button).find(".guessBtn");
  expect(guessButtons).toHaveLength(4);

  // TODO: check texts
});
