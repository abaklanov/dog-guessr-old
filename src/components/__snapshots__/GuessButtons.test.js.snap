// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="row"
    key="0"
  >
    <div
      className="col-lg-4"
    >
      <ForwardRef(Bootstrap(Button))
        className="guessBtn"
        variant="light"
      >
        dingo
      </ForwardRef(Bootstrap(Button))>
    </div>
  </div>
  <div
    className="row"
    key="1"
  >
    <div
      className="col-lg-4"
    >
      <ForwardRef(Bootstrap(Button))
        className="guessBtn"
        variant="light"
      >
        cattledog
      </ForwardRef(Bootstrap(Button))>
    </div>
  </div>
  <div
    className="row"
    key="2"
  >
    <div
      className="col-lg-4"
    >
      <ForwardRef(Bootstrap(Button))
        className="guessBtn"
        variant="light"
      >
        cattledog
      </ForwardRef(Bootstrap(Button))>
    </div>
  </div>
  <div
    className="row"
    key="3"
  >
    <div
      className="col-lg-4"
    >
      <ForwardRef(Bootstrap(Button))
        className="guessBtn"
        variant="light"
      />
    </div>
  </div>
</div>
`;
